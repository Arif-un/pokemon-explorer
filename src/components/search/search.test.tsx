import { type DependencyList } from 'react'
import { useDebounce } from 'react-use'

import { useNavigate, useSearch } from '@tanstack/react-router'
import { act, fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import Search from './search'

vi.mock('@tanstack/react-router', () => ({
  useNavigate: vi.fn(),
  useSearch: vi.fn()
}))

vi.mock('react-use', () => ({
  useDebounce: vi.fn()
}))

describe('Search', () => {
  const mockNavigate = vi.fn()
  let mockSetDebouncedValue: (value: string) => void

  beforeEach(() => {
    vi.mocked(useNavigate).mockReturnValue(mockNavigate)
    vi.mocked(useSearch).mockReturnValue({ name: '' })
    vi.mocked(useDebounce).mockImplementation(
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
      (callback: Function, _delay: number, _deps: DependencyList) => {
        mockSetDebouncedValue = callback as (value: string) => void
      }
    )
    vi.clearAllMocks()
  })

  it('renders correctly', () => {
    const { asFragment } = render(<Search />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('displays the search input with correct placeholder', () => {
    render(<Search />)
    const input = screen.getByPlaceholderText('Search for a Pokémon...')
    expect(input).toBeInTheDocument()
  })

  it('updates search query on input change', () => {
    render(<Search />)
    const input = screen.getByPlaceholderText('Search for a Pokémon...') as HTMLInputElement
    fireEvent.change(input, { target: { value: 'Pikachu' } })
    expect(input.value).toBe('Pikachu')
  })

  it('calls navigate when debounced value changes', () => {
    render(<Search />)
    const input = screen.getByPlaceholderText('Search for a Pokémon...')
    fireEvent.change(input, { target: { value: 'Charizard' } })

    act(() => {
      mockSetDebouncedValue('Charizard')
    })

    expect(mockNavigate).toHaveBeenCalledWith({
      search: expect.any(Function),
      resetScroll: false
    })
  })

  it('initializes search query from URL parameters', () => {
    vi.mocked(useSearch).mockReturnValue({ name: 'Bulbasaur' })
    render(<Search />)
    const input = screen.getByPlaceholderText('Search for a Pokémon...') as HTMLInputElement
    expect(input.value).toBe('Bulbasaur')
  })

  it('does not navigate on first render with empty query', () => {
    render(<Search />)
    expect(mockNavigate).not.toHaveBeenCalled()
  })
})
