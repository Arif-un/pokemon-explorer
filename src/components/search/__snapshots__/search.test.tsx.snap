// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="relative focus-within:text-yellow-500 dark:text-white inline-block"
  >
    <input
      class="h-11 inline-block rounded-full px-4 w-64 pl-10 border border-gray-300 dark:border-gray-700 bg-slate-50 dark:bg-slate-900 text-slate-950 dark:text-white focus:outline-none focus:ring-2 ring-yellow-500  focus:bg-white focus:dark:bg-slate-800"
      placeholder="Search for a PokÃ©mon..."
      type="search"
      value=""
    />
    <svg
      class="lucide lucide-search absolute inset-y-0 my-auto left-4"
      fill="none"
      height="18"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="18"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="11"
        cy="11"
        r="8"
      />
      <path
        d="m21 21-4.3-4.3"
      />
    </svg>
  </div>
</DocumentFragment>
`;
